### PR TITLE
Support RV32E

### PR DESCRIPTION
**Pull request by kito-cheng**
Open a new PR for RV32E :P

### CODE REVIEW

File: gas/config/tc-riscv.c [hunk 1]
```diff
@@ -171,6 +182,13 @@ riscv_set_arch (const char *s)
       case 'i':
 	break;
 
+      case 'e':
+	p++;
+	riscv_add_subset ("e");
+	riscv_add_subset ("i");
+	riscv_set_rve (TRUE);
```
Comment (aswaterman) [inline gas/config/tc-riscv.c]: I think we want to add

    if (xlen > 32)
      as_fatal ("-march=%s: rv%de is not a valid base ISA", s, xlen);

**Comment by aswaterman** [comment]
I think we should merge https://github.com/riscv/riscv-binutils-gdb/pull/102 first, then add a similar check to enforce prerequisites:

    if (riscv_subset_supports ("e") && riscv_subset_supports ("f"))
      as_fatal ("-march=%s: rv32e does not support the `f' extension", s);

**Comment by aswaterman** [comment]
Also, note that RV32E Is only version 1.9, so there may be some changes before it is frozen.  In particular, the C extension might be slightly different for RVE (e.g., ```c.lbu``` and ```c.lh``` might replace ```c.flw``` and ```c.fld```).

I think we can merge this patch anyway--just a warning.

**Comment by kito-cheng** [comment]
> Also, note that RV32E Is only version 1.9, so there may be some changes before it is frozen. In particular, the C extension might be slightly different for RVE (e.g., c.lbu and c.lh might replace c.flw and c.fld).

Ya, we are considering to recycle those encoding space too.

File: gas/config/tc-riscv.c (hunk 1 above)
Comment (kito-cheng) [inline gas/config/tc-riscv.c]: Done.

**Comment by kito-cheng** [comment]
Rebase to #102,

**Comment by palmer-dabbelt** [comment]
I'll submit this upstream in a bit, everything's a bit of a mess right now...

**Comment by kito-cheng** [comment]
I guess maybe I can back to support the gcc part next month, we are mess in several meaning too...:P

**Pull request merged**